### PR TITLE
refactor and restructuring of sampler code to give custom nodes more access

### PR DESCRIPTION
**Pull request by BlenderNeko**
This PR makes the following changes:

- refactors the code in [common_ksampler](https://github.com/comfyanonymous/ComfyUI/blob/master/nodes.py#L743) to a set of functions that are easier to reuse.
- makes it possible for other code to calculate sigmas without creating a Ksampler to call [_calculate_sigmas](https://github.com/comfyanonymous/ComfyUI/blob/master/comfy/samplers.py#L432)
- allow calls to [Ksampler.sample](https://github.com/comfyanonymous/ComfyUI/blob/master/comfy/samplers.py#L465) to supply a set of sigmas to be used instead of the pre-calculated ones.

I was not entirely sure where to put the functions that resulted from the refactoring of the common_ksampler function, these currently reside in comfy/sample.py

These changes make it possible to for instance create the custom nodes in [this](https://github.com/BlenderNeko/ComfyUI_Noise) repo

### CODE REVIEW

File: comfy/sample.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import torch
+import comfy.model_management
+
+
+def prepare_noise(latent, seed):
+    """creates random noise given a LATENT and a seed"""
+    latent_image = latent["samples"]
+    batch_index = 0
+    if "batch_index" in latent:
+        batch_index = latent["batch_index"]
```
Comment (comfyanonymous) [inline comfy/sample.py]: I feel like the stuff in the comfy folder shouldn't know about the internals of the latent dict. The contents should be passed as params to the function.

Comment (BlenderNeko) [reply comfy/sample.py]: I made the params in the function calls more explicit. I still need to see if I didn't accidentally broke any of the more advanced stuff (masks, gligen, controlnets).

Comment (BlenderNeko) [reply comfy/sample.py]: Tested on inpainting, control nets, selected batch index and gligen. All features seem to be working as intended.

**Comment by comfyanonymous** [comment]
I moved the calculate_sigmas back to the KSampler object because I think it makes the code too fragile to have it somewhere else. You can just instantiate a KSampler if you want to call it.
